packages/modelence/src/auth/providers/google.ts

```diff
@@ -0,0 +1,103 @@
+import { Router, type Request, type Response } from "express";
+import { ObjectId } from "mongodb";
+import passport from "passport";
+import { Strategy as GoogleStrategy } from "passport-google-oauth20";
+import { usersCollection } from "../db";
+import { createSession } from "../session";
+
+interface GoogleUser {
+  id: string;
+  displayName: string;
+  email: string;
+  photos: { value: string }[];
+}
+
+async function authenticateUser(res: Response, userId: ObjectId) {
+  const { authToken } = await createSession(userId);
+
+  res.cookie("authToken", authToken, {
+    httpOnly: true,
+    secure: process.env.NODE_ENV === "production",
+    sameSite: "strict",
+  });
+  res.status(301);
+  res.redirect("/");
+}
+
+
+async function handleGoogleAuthenticationCallback(req: Request, res: Response) {
+  const googleUser = req.user as GoogleUser;
+
+  const existingUser = await usersCollection.findOne(
+    { 'google.id': googleUser.id },
+  );
+
+  if (existingUser) {
+    await authenticateUser(res, existingUser._id);
+    
+    return;
+  }
+
+  const existingUserByEmail = await usersCollection.findOne(
+    { 'emails.address': googleUser.email, },
+    { collation: { locale: 'en', strength: 2 } },
+  );
+
+  // TODO: check if the email is verified
+  if (existingUserByEmail) {
+    // If the user exists but has a different Google ID, we can link the Google account
```

Comment:
Can we just throw an error in this case? We used to do it that way at CodeSignal, because we don't know for sure if it's the same user. Also, I may have somehow connected my X@gmail.com google account to a user account with my Y@gmail.com email address, so we'll be overriding it.

I believe the right way to link is to explicitly prompt the user to link their existing account:
- User logs in via Google
- We detect the email is already registered, ask them if they want to merge
- If they say yes, they log in with email/password as the next step to their existing account
- Google account is not linked to the existing account

But since this proper flow is a little complicated we can just omit it for now and throw an error instead, saying sign up with password instead.